### PR TITLE
Import local libs as extensions

### PR DESCRIPTION
**Pull request by Fr6jDJF**
**Basic loader for local libraries with optional looping tasks**.

- Custom libraries placed in the  _local-lib_ folder(configuration.default.ini) can be installed as in example "local_lib_pipinstal.sh" with:
venv/bin/python -m pip install -e local-lib/testlib
As it will be installed as *.egg-link in venv/../../site-packages, updating is a matter of updating the files in local-lib folder.
- list with libs to be loaded in configuration.default.ini
- lib has it's own libconfiguration.ini
- lib has it's own lang translation files *example lib holds help message(!test command only) that extend the main help string.
- No lib reload(just restart bot)
- No auto dependency install(TestLib performs one basic check on it's own for _feedparser_ example), need to figure out if/how to have those installed by the setup files

### CODE REVIEW

**Comment by TerryGeng** [comment]
Hello. Thanks for contributing to our project.

I have noticed you made a lot of changes. Can you provide details about the functionalities you'd like to add? I guess what you are aiming for is adding additional commands and hooks in the main loop of the bot by adding "libraries". Did I get it right?

File: local_lib_pipinstal.sh
Comment (TerryGeng) [inline local_lib_pipinstal.sh:5]: We don't use pip to manage our project, since @azlux has his own auto-update engine. I think it would be messy if we choose to use pip to manage the lib install...

File: mumbleBot.py
Comment (TerryGeng) [inline mumbleBot.py:36]: Please keep this line. Azlux has the routine for replacing this line with the correct version when doing a release. The 'git' stands for the development version and will be dealt with https://github.com/azlux/botamusique/blob/014086305d54a2b585839f34e8d279c384658596/mumbleBot.py#L209.

File: local-lib/testlib/testlib/typical.py
Comment (TerryGeng) [inline local-lib/testlib/testlib/typical.py:26]: I think this Lib does something like reading RSS feed. Can you elaborate more? And if this Lib does do some thing, then we should give it a meaningful name like RSSLib.

File: local-lib/testlib/testlib/typical.py
Comment (TerryGeng) [inline local-lib/testlib/testlib/typical.py:81]: load_mod.

File: mumbleBot.py
Comment (TerryGeng) [inline mumbleBot.py:510]: I happen to be in the middle of considering optimizing the logic of the main loop to make it lighter. And now we have these dict lookups in the middle...

I think if you create a specific class like
```python
from enum import Enum

class TaskState(Enum):
    added = 0
    started = 1

class TaskHandle:
    def __init__(self, name, state, handle):
        self.name = name
        self.state = TaskState.added
        self.handle = handle
    ...
```

However, I'm just wondering why you put task initialization in the main loop?

File: mumbleBot.py
Comment (TerryGeng) [inline mumbleBot.py:928]: import_all_local_libs

**Review by TerryGeng — CHANGES_REQUESTED** [review]
Sorry that I don't have time to do a full review. I just briefly glanced through your changes and posted some comments.

I think we still need to do a lot of clean up and functionalities tests before I can merge it.

Actually, I'd appreciate if you could start an issue and discuss your plan with us before you get started...

@azlux and I are always on IRC channel #mumble at freenode.net and feel free to drop by and have some small discussion with us :)

**Comment by Fr6jDJF** [comment]
> Hello. Thanks for contributing to our project.
> 
> I have noticed you made a lot of changes. Can you provide details about the functionalities you'd like to add? I guess what you are aiming for is adding additional commands and hooks in the main loop of the bot by adding "libraries". Did I get it right?
Correct, 
For me, it's non-general functions that interact with running services(for now I have  some broadcast info to chat), but it might just as well extend to more general libs supporting the discussions on Mumble, like IMDB look ups, game server status and such.

So a.t.m I manually inject my own code to interact with my running services, what means auto update must be disabled, aging the instance. functionality like this would be the solution for that.

File: mumbleBot.py
Comment (Fr6jDJF) [reply mumbleBot.py:510]: Maybe It's better to just release the loop of already started thread tasks in main, this would also give better control of stopping/pause the task.
I currently placed it in the loop the get them running as late as possible, as they live quite on their own.

File: local-lib/testlib/testlib/typical.py
Comment (Fr6jDJF) [reply local-lib/testlib/testlib/typical.py:26]: This lib was pure for example/testing, I could bring it down as suggested being the example RSSLib

File: local_lib_pipinstal.sh
Comment (Fr6jDJF) [reply local_lib_pipinstal.sh:5]: I was unaware of this (auto update disabled by default).

File: mumbleBot.py
Comment (TerryGeng) [inline mumbleBot.py:209]: Can you explain this deletion?

File: mumbleBot.py
Comment (TerryGeng) [inline mumbleBot.py:657]: Can you explain this part?

Logics related to the loop and its state are messy and modifications, if not thoroughly tested, if prone to errors. Are you sure this will work?

Comment (TerryGeng) [reply mumbleBot.py:657]: As you can see, this part was introduced in https://github.com/azlux/botamusique/commit/fceb50e8943fb1f92537cd59b3b2ed16c14ce29b, to fix a bug. I think you need to keep the part unchanged.

**Comment by TerryGeng** [comment]
Great! Thanks for the improvement.

I'd suggest you check out the ABC of python, and write a template ABC class for all Libs inside libimport.py (or I think you can rename libs into plugins and rename libimport.py into plugins.py to avoid confusion).

Another thing: I think you don't have to use force push in this PR. Using force push makes it hard for me to find out what is your improvement across each commit. You can just commit as usual and we will squash this PR before merging. What's your idea?

File: mumbleBot.py
Comment (TerryGeng) [inline mumbleBot.py:519]: What do you think of moving this part to the place between L516 and 517, outside the while loop?

File: mumbleBot.py
Comment (TerryGeng) [inline mumbleBot.py:524]: I'm kind of confused about "started", "released" and "stopped". Can you explain a little bit?

File: mumbleBot.py
Comment (Fr6jDJF) [reply mumbleBot.py:519]: That would be better as there is no active monitoring by the bot.

File: mumbleBot.py
Comment (Fr6jDJF) [reply mumbleBot.py:524]: added: init
started: thread is running
released: task may start/resume it's job
paused: task waiting for released
stopped: exit/break task loop

Any renaming suggestions, or would you rather see just threads being released to do their stuff?

As for the interaction with the threads maybe [queue](https://docs.python.org/3.8/library/queue.html) is something to consider here, and feed the q object some addressed TaskHandle , 
For now I will leave it up to further discussion.

**Comment by Fr6jDJF** [comment]
> Great! Thanks for the improvement.
> 
> I'd suggest you check out the ABC of python, and write a template ABC class for all Libs inside libimport.py 

I'm not known with that, I will check that out, thank you.
>(or I think you can rename libs into plugins and rename libimport.py into plugins.py to avoid confusion).

Ok, that would be:


                        shown with log.debug
                                 v

/local-lib/rssplugin/rssplugin/typ_rssplugin.Rss

            ^_________^
       would translate(shown by pip freeze) in to "local-rssplugin" by pip setup.cfg


> Another thing: I think you don't have to use force push in this PR. Using force push makes it hard for me to find out what is your improvement across each commit. You can just commit as usual and we will squash this PR before merging. What's your idea?

I know sorry, I'm not used to push PR, to me -f seemed the only way to get this PR updated again.

**Comment by TerryGeng** [comment]
> I know sorry, I'm not used to push PR, to me -f seemed the only way to get this PR updated again.

Actually you can just keep pushing commits into your 'importlib' branch, and new commits will just show up here :).

**Review by Fr6jDJF — COMMENTED** [review]
updating

**Comment by TerryGeng** [comment]
Hello! Any update on this PR?

**Comment by luca0N** [comment]
> Hello! Any update on this PR?

Looks like this user hasn't been active on GitHub since January 3, 2021... I would love to continue to work on this PR to get it merged on botamusique, though.

@TerryGeng, could we get a checklist going to see what's left for this to get merged?

I will look into this as soon as I have some time.

**Pull request closed without merging**
